File: api/cpp/example/kv_example.cpp
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2023-2025. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <iostream>

#include "yr/yr.h"

int main()
{
    YR::Config conf;
    conf.functionUrn = "sn:cn:yrk:12345678901234561234567890123456:function:0-opc-opc:$latest";
    conf.serverAddr = "";      // bus proxy IP:port
    conf.dataSystemAddr = "";  // datasystem worker IP:port
    std::vector<std::string> paths;
    paths.emplace_back("");  // add your library path
    conf.loadPaths = std::move(paths);
    YR::Init(conf);

    std::string key = "kv-id-888";
    std::string value = "kv-value-888";
    YR::KV().Write(key, value);

    std::shared_ptr<std::string> result = YR::KV().Read<std::string>(key);
    std::cout << *result << std::endl;

    YR::KV().Del(key);

    // Legacy API
    YR::KV().Set(key, value);

    std::string result2 = YR::KV().Get(key);
    std::cout << result2 << std::endl;

    YR::KV().Del(key);

    // Set with SetParam
    YR::ExistenceOpt existence = YR::ExistenceOpt::NONE;
    YR::WriteMode writeMode = YR::WriteMode::NONE_L2_CACHE;
    YR::SetParam setParam = {
        .writeMode = writeMode,
        .ttlSecond = 5,
        .existence = existence,
    };
    YR::KV().Set(key, value, setParam);

    std::string result3 = YR::KV().Get(key);
    std::cout << result3 << std::endl;
    sleep(8);  // sleep 8s
    try {
        std::string result4 = YR::KV().Get(key, 10);
        std::cout << result4 << std::endl;
    } catch (YR::Exception &e) {
        std::cout << e.what() << std::endl;
    }

    YR::Finalize();
    return 0;
}

File: api/go/example/kv_example.go
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2025. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package main

import (
	"flag"
	"fmt"
	"os"
	"os/signal"
	"syscall"

	"yuanrong.org/kernel/runtime/yr"
)

func InitYR() {
	flag.Parse()
	config := &yr.Config{
		Mode:           yr.ClusterMode,
		FunctionUrn:    "sn:cn:yrk:default:function:0-opc-opc:$latest",
		ServerAddr:     flag.Args()[0],
		DataSystemAddr: flag.Args()[1],
		InCluster:      true,
		LogLevel:       "DEBUG",
	}
	info, err := yr.Init(config)
	if err != nil {
		fmt.Println("Init failed, error: ", err)
	}
	fmt.Println(info)
}

func KVSetAndGetAndDelExample() {
	err := yr.SetKV("key", "value")
	if err != nil {
		fmt.Println("set kv failed, error: ", err)
	}

	fmt.Println(yr.GetKV("key", 30000))

	fmt.Println(yr.DelKV("key"))
}

func BatchKVSetAndGetAndDelExample() {
	err := yr.SetKV("key", "value")
	if err != nil {
		fmt.Println("set kv failed, error: ", err)
	}

	err = yr.SetKV("key1", "value1")
	if err != nil {
		fmt.Println("set kv failed, error: ", err)
	}

	fmt.Println(yr.GetKVs([]string{"key", "key1"}, 30000, false))

	fmt.Println(yr.DelKVs([]string{"key", "key1"}))

	fmt.Println(yr.GetKVs([]string{"key", "key1"}, 30000, false))
}

// compile command:
// CGO_LDFLAGS="-L$(path which contains libcpplibruntime.so)" go build -o main kv_example.go
func main() {
	c := make(chan os.Signal)
	signal.Notify(c, syscall.SIGTERM, syscall.SIGKILL, syscall.SIGINT)
	InitYR()
	KVSetAndGetAndDelExample()
	BatchKVSetAndGetAndDelExample()
	<-c
}
